lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+    -- coins, sorted into an order that ensures all empty token maps are at the
+    -- start of the list.
+    -> Maybe [TokenBundle]
+    -- ^ Resulting change bundles, or 'Nothing' if there was not enough ada
+    -- available to assign a minimum ada quantity to all non-empty token maps.
+assignCoinsToChangeMaps adaAvailable minCoinFor pairsAtStart =
+    loop adaRequiredAtStart pairsAtStart
+  where
+    loop !adaRequired !pairsNonEmpty = case pairsNonEmpty of
+
+        pair :| pairs | adaAvailable >= adaRequired ->
+            -- We have enough ada available to pay for the minimum required
+            -- amount of every asset map that remains in our list:
+            let assetMapsRemaining = fst <$> (pair :| pairs) in
+            let bundlesForAssetsWithMinimumCoins =
+                    assignMinimumCoin minCoinFor <$> assetMapsRemaining in
+            -- Calculate the amount of ada that remains after assigning the
+            -- minimum amount to each map. This should be safe, as we have
+            -- already determined that we have enough ada available:
+            let adaRemaining = adaAvailable `Coin.distance` adaRequired in
+            -- Partition any remaining ada according to the weighted
+            -- distribution of output coins that remain in our list:
+            let outputCoinsRemaining = snd <$> (pair :| pairs) in
+            let bundlesForOutputCoins = TokenBundle.fromCoin <$>
+                    makeChangeForCoin outputCoinsRemaining adaRemaining in
+            -- Finally, combine the minimal coin asset bundles with the
+            -- bundles obtained by partitioning the remaining ada amount:
+            Just $ NE.toList $ NE.zipWith (<>)
+                bundlesForAssetsWithMinimumCoins
+                bundlesForOutputCoins
+
+        (m, _) :| (p : ps) | TokenMap.isEmpty m && adaAvailable < adaRequired ->
+            -- We don't have enough ada available to pay for the minimum
+            -- required amount of every asset map, but we do have an empty
+            -- asset map that is safe to drop. This will reduce the amount of
+            -- ada required by a small amount:
+            let adaRequired' = adaRequired `Coin.distance` minCoinFor m in
+            loop adaRequired' (p :| ps)
+
+        (m, _) :| [] | TokenMap.isEmpty m && adaAvailable >= adaRequired ->
+            -- We didn't have any non-ada assets at all in our change. We can
+            -- simply return the available ada amount as a singleton change
+            -- bundle:
+            Just [TokenBundle.fromCoin adaAvailable]
+
+        (m, _) :| [] | TokenMap.isEmpty m && adaAvailable < adaRequired ->
+            -- We didn't have any non-ada assets at all in our change, but we
+            -- also don't have enough ada available to pay even for a single
+            -- change output. We just burn the available ada amount (which
+            -- will be small), returning no change.
+            Just []
```

Comment:
Yes. I like that is guarded by `adaAvailable < adaRequired`. Although I'd really prefer that the `adaRequired` is actually computed within the function loop itself, and not provided as argument. It makes reasoning and following the steps much easier and it becomes clear that `adaRequired` is necessarily small.

Reply:
Hi @KtorZ 

I've written a justification for this in https://github.com/input-output-hk/cardano-wallet/pull/2490#discussion_r570169560. Would be interested to know your thoughts on that!